### PR TITLE
Convert patterns to ghc-lib

### PR DESCRIPTION
**Pull request by shayne-fletcher-da**
Reimplement `Pattern.hs` to use GHC parse trees.

### CODE REVIEW

File: src/Hint/Pattern.hs [hunk 1]
```diff
@@ -118,81 +102,133 @@ hints gen (Pattern pats (GuardedRhss _ [GuardedRhs _ [Generator _ pat (App _ op
...
-patHint False strict o@(PIrrPat _ x) | f x = [warn "Redundant irrefutable pattern" o x [r]]
-    where f (PParen _ x) = f x
-          f (PAsPat _ _ x) = f x
-          f PWildCard{} = True
-          f PVar{} = True
-          f _ = False
-          r = Replace R.Pattern (toSS o) [("x", toSS x)] "x"
-patHint lang strict o@(PAsPat u v PWildCard{}) = [warn "Redundant as-pattern" o v []]
+asPattern :: LHsDecl GhcPs  -> [(Pattern, String -> Pattern -> [Refactoring R.SrcSpan] -> Idea)]
+asPattern (LL loc x) = concatMap decl (universeBi x) -- ++ concatMap alt (universeBi x)
```
Comment (googleson78) [inline src/Hint/Pattern.hs]: Commented out code left here?

Comment (shayne-fletcher-da) [reply src/Hint/Pattern.hs]: Nice catch.

File: src/GHC/Util/HsPat.hs [hunk 2]
```diff
@@ -1,9 +1,9 @@
 {-# LANGUAGE MultiParamTypeClasses , FlexibleInstances, FlexibleContexts #-}
```
Comment (ndmitchell) [inline src/GHC/Util/HsPat.hs]: Why the rename?

**Comment by ndmitchell** [comment]
Thanks!

File: src/GHC/Util/HsPat.hs (hunk 2 above)
Comment (shayne-fletcher-da) [reply src/GHC/Util/HsPat.hs]: Not a technical reason. The naming scheme for these modules is that they are named after the (module/)datatype for which they relate e.g. `GHC.Util.HsExpr` for `HsExpr`, `GHC.Util.HsDecl` for `HsDecl` and so on.  Actually, reviewing this I see I that in fact, though the module name is `HsPat`, the type is now `Pat` (doh!) so it was right the way it was. I'll revert the rename.

All the above said, don't know if you are aware - on HEAD there is lots of churn here to introduce a new "GHC module hierarchy" and everything is getting moved.

**Pull request merged**